File: solutions/haiku-1/haiku/typescript/roman.ts
import { readFileSync } from 'fs';

const romanValues = [
  { value: 1000, numeral: 'M' },
  { value: 900, numeral: 'CM' },
  { value: 500, numeral: 'D' },
  { value: 400, numeral: 'CD' },
  { value: 100, numeral: 'C' },
  { value: 90, numeral: 'XC' },
  { value: 50, numeral: 'L' },
  { value: 40, numeral: 'XL' },
  { value: 10, numeral: 'X' },
  { value: 9, numeral: 'IX' },
  { value: 5, numeral: 'V' },
  { value: 4, numeral: 'IV' },
  { value: 1, numeral: 'I' },
];

function intToRoman(num: number): string {
  let result = '';
  for (const { value, numeral } of romanValues) {
    while (num >= value) {
      result += numeral;
      num -= value;
    }
  }
  return result;
}

const input = readFileSync(0, 'utf-8');
const lines = input.trim().split('\n');
for (const line of lines) {
  const num = parseInt(line, 10);
  if (num > 0) {
    console.log(intToRoman(num));
  }
}

File: solutions/haiku-1/haiku/rust/roman.rs
use std::io::{self, BufRead};

fn main() {
    let stdin = io::stdin();
    let reader = stdin.lock();

    for line in reader.lines() {
        if let Ok(line) = line {
            let num: u32 = line.trim().parse().unwrap();
            println!("{}", int_to_roman(num));
        }
    }
}

fn int_to_roman(mut num: u32) -> String {
    let values = [
        (1000, "M"),
        (900, "CM"),
        (500, "D"),
        (400, "CD"),
        (100, "C"),
        (90, "XC"),
        (50, "L"),
        (40, "XL"),
        (10, "X"),
        (9, "IX"),
        (5, "V"),
        (4, "IV"),
        (1, "I"),
    ];

    let mut result = String::new();

    for &(value, symbol) in &values {
        while num >= value {
            result.push_str(symbol);
            num -= value;
        }
    }

    result
}
